src/lib/reasoners/adt_rel.ml

```diff
@@ -439,6 +441,8 @@ let assume_literals la uf env =
 
        | Distinct (false, [r1; r2]) as l, _, ex, _ when is_adt r2 ->
          Debug.assume l;
+         let r1, _ = Uf.find_r uf r1 in
+         let r2, _ = Uf.find_r uf r2 in
```

Comment:
The explanations should be added to `ex`

Reply:
Uhm, I wasn't sure it was necessary because the explanation may be added by a `subst` literal? I agree it's safer to add it.

Reply:
It is necessary. If `ex` justifies that `r1 <> r2`, `ex` on its own is not enough to justify `rr1 <> rr2`.

(For instance if we have `x <> y` and `x = x'` ; `y = y'`, `x <> y` alone is not sufficient to justify `x' <> y'` ; we need all 3 facts (`x <> y`, `x = x'` and `y = y'`) to derive `x' <> y'`)

Reply:
Note that there is no guarantee about the order in which we treat the literals (they are shuffled in `make_unique`).